### PR TITLE
Adding go 1.21 devfile

### PR DESCRIPTION
**Pull request by ibuziuk**
### What does this PR do?:
Adding go 1.21 devfile

### Which issue(s) this PR fixes:\
N/A

### PR acceptance criteria:

- [x] Contributing guide
_Have you read the [devfile registry contributing guide](https://github.com/devfile/registry/blob/main/CONTRIBUTING.md) and followed its instructions?_
- [x] Test automation
_Does this repository's tests pass with your changes?_
- [ ] Documentation
_Does any documentation need to be updated with your changes?_
- [x] Check Tools Provider
_Have you tested the changes with existing tools, i.e. Odo, Che, Console? (See [devfile registry contributing guide](https://github.com/devfile/registry/blob/main/CONTRIBUTING.md) on how to test changes)_

### How to test changes / Special notes to the reviewer:

https://workspaces.openshift.com#https://raw.githubusercontent.com/devfile/registry/747c75aeefb71de0f154c77cf307940a9b7069c1/stacks/go/2.4.0/devfile.yaml

### CODE REVIEW

**Review by thepetk — COMMENTED** [review]
@ibuziuk in general looks good to me, one minor suggestion is to add also a `1.4.0` version using `go 1.21` & have an `innerloop` approach.

**Comment by ibuziuk** [comment]
@thepetk thanks for the review, but I'm not a fan of the versioning approach in general. Could you clarify what is the difference between go `1.x.x` and `2.x.x` and why do we need both

**Comment by thepetk** [comment]
> @thepetk thanks for the review, but I'm not a fan of the versioning approach in general. Could you clarify what is the difference between go `1.x.x` and `2.x.x` and why do we need both

So far the `1.x.x` versions support `innerloop` scenarios and the `2.x.x` `outerloop` scenarios.

**Comment by ibuziuk** [comment]
@thepetk any reason why not to have one devfile that combines both scenarios?

**Comment by thepetk** [comment]
> @thepetk any reason why not to have one devfile that combines both scenarios?

I think we already combine both scenarios. From my perspective, your current approach of the `2.4.0` already covers both approaches (like most of the 2.x.x approaches). My point for "backporting" the update to go 1.21 also in `1.x.x` versions, is that the current default version of the `go` stack is `1.x.x`. So IMO creating the `1.4.0` stack version could be good addition.

**Comment by ibuziuk** [comment]
@thepetk @michael-valdron folks, I updated the PR based on your feedback, but I do not really understand what you are trying to achieve with versioning. Yes, I do understand that sometimes there could be compatibility reasons, but this does not seem to be the case. Why do you want to continue having 1.x / 2.x versions pattern of Go devfiles. For me personally it is super confusing. I would even understand if you would like to have identical devfiles with only go version diff e.g. 1.19, 1.20, 1.21 etc. but this inner / outer loop approach with 1.x  / 2.x does not make any sense imo

File: stacks/go/1.4.0/devfile.yaml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+schemaVersion: 2.2.2
```
Comment (michael-valdron) [inline stacks/go/1.4.0/devfile.yaml:1]: 1.x stack versions should be still using devfile 2.1.0:
```suggestion
schemaVersion: 2.1.0
```

**Review by michael-valdron — CHANGES_REQUESTED** [review]
In addition to my comment, I'd also recommend deprecating the patch versions introduced by https://github.com/devfile/registry/pull/357 to prompt anyone using these versions to either bump to `1.4.0` or switch to `2.x` versions if they need to use the `2.2.x` devfile specification.

**Comment by michael-valdron** [comment]
> @thepetk @michael-valdron folks, I updated the PR based on your feedback, but I do not really understand what you are trying to achieve with versioning. Yes, I do understand that sometimes there could be compatibility reasons, but this does not seem to be the case. Why do you want to continue having 1.x / 2.x versions pattern of Go devfiles. For me personally it is super confusing. I would even understand if you would like to have identical devfiles with only go version diff e.g. 1.19, 1.20, 1.21 etc. but this inner / outer loop approach with 1.x / 2.x does not make any sense imo

Difference of separating 1.x / 2.x is that some tools don't have compatibility with the 2.2.x devfile spec, such as odo v2, this is why 1.x was left with just an inner loop setup and 2.x includes both. Even if a tool only supports inner loop the tool _should_ still use 2.x version of the stack if the tool supports the 2.2.x devfile spec.

The change merged in https://github.com/devfile/registry/pull/357 however has bumped the schema version for 1.x to 2.2.x devfile spec, an oversight on our part. I would recommend [reverting the 1.x stack back to using devfile 2.1.0](https://github.com/devfile/registry/pull/386#discussion_r1605251805) to maintain the purpose of this stack version and if there are use cases need to use the 2.2.x specification to switch to the 2.x stack instead.

Apologies for the mix up and misunderstanding here.

cc @ibuziuk @thepetk

File: stacks/go/1.4.0/devfile.yaml (hunk 1 above)
Comment (ibuziuk) [reply stacks/go/1.4.0/devfile.yaml:1]: why?

Comment (michael-valdron) [reply stacks/go/1.4.0/devfile.yaml]: https://github.com/devfile/registry/pull/386#issuecomment-2117944019

**Comment by ibuziuk** [comment]
@michael-valdron I removed 1.x version from the PR. Feel free to merge it in the current state or close since I do not plan to provide any more updates

**Comment by michael-valdron** [comment]
Using https://github.com/devfile/api/issues/1564 to track these changes towards the v1 go stack.

> The change merged in https://github.com/devfile/registry/pull/357 however has bumped the schema version for 1.x to 2.2.x devfile spec, an oversight on our part. I would recommend https://github.com/devfile/registry/pull/386#discussion_r1605251805 to maintain the purpose of this stack version and if there are use cases need to use the 2.2.x specification to switch to the 2.x stack instead.

This issue is being by https://github.com/devfile/api/issues/1563 and will be handled by us.

**Review by michael-valdron — COMMENTED** [review]
@ibuziuk Changes looks good to me, need a sign off from someone on your team then I can sign off on this.

cc @devfile/che-team

File: stacks/go/2.4.0/devfile.yaml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        rootRequired: false
+  - name: deploy
+    kubernetes:
+      uri: kubernetes/deploy.yaml
+      endpoints:
+      - name: http-8081
+        targetPort: 8081
+  - container:
+      endpoints:
+        - name: http-go
```
Comment (svor) [inline stacks/go/2.4.0/devfile.yaml:37]:
```suggestion
        - name: https-go
          protocol: https
```

**Review by svor — COMMENTED** [review]
I was able to start a workspace on Che, but commands are failed:
screenshot-devspaces apps sandbox-m4 g2pi p1 openshiftapps com-2024 05 20-10_24_52

I think we have an issue for that: https://github.com/eclipse-che/che/issues/22930

Also if we want to see this version on the Che Dashboard, need to mark it as a default one in stack.yaml file

**Comment by svor** [comment]
@tolusha am I right that it's a known issue that commands from Go sample are failing on Che

**Comment by tolusha** [comment]
@svor 
My guess, this issue [1] is mostly about nonsupporting env vars in the commands. At least it looks like.
As a result, the following exec command fails:
```yaml
 exec:
      env:
        - name: GOPATH
          value: ${PROJECT_SOURCE}/.go
        - name: GOCACHE
          value: ${PROJECT_SOURCE}/.cache
      commandLine: go build main.go
```

[1] https://github.com/eclipse-che/che/issues/22930

**Review by svor — APPROVED** [review]
lgtm

**Pull request merged**